Byte Array Interop Breaking Change Announcement (#24759)
dotnet/docs: docs/core/compatibility/aspnet-core/6.0/byte-array-interop.md

---
title: "Breaking change: Blazor: Byte Array Interop"
description: "Learn about the breaking change in ASP.NET Core 6.0 titled Blazor: Byte Array Interop"
no-loc: [ Blazor ]

> ms.author: tanayparikh

Reviewer: ```suggestion
ms.author: taparikh
```

This one has to be a Microsoft alias.

ms.date: 06/21/2021
---
# Blazor: Byte-array interop

> Blazor now supports optimized byte array interop which avoids encoding and decoding byte-arrays into Base64 and facilitates a more efficient interop process. This applies to both Blazor Server and Blazor WebAssembly.

Reviewer: ```suggestion
Blazor now supports optimized byte-array interop, which avoids encoding and decoding byte-arrays into Base64 and facilitates a more efficient interop process. This applies to both Blazor Server and Blazor WebAssembly.
```

---
title: "Breaking change: Blazor: Byte Array Interop"
description: "Learn about the breaking change in ASP.NET Core 6.0 titled Blazor: Byte Array Interop"
no-loc: [ Blazor ]
ms.author: tanayparikh
ms.date: 06/21/2021
---
# Blazor: Byte Array Interop

> Blazor Server & WebAssembly now supports optimized byte array interop which avoids encoding/decoding byte arrays into Base64, facilitating a more efficient interop process.

Reviewer 2: ```suggestion
Blazor now supports optimized byte array interop which avoids encoding/decoding byte arrays into Base64, facilitating a more efficient interop process. This applies to both Blazor Server and Blazor WebAssembly.
```

## Version introduced

ASP.NET Core 6.0 Preview 6

## Returning byte array from JS to .NET

### Old behavior

> ```csharp

Reviewer 2: ```suggestion
```typescript
```

This is not C#, and it actually looks like TypeScript since there is an `as` cast.

function someJSMethodReturningAByteArray() {
    const data = new Uint8Array([ 1, 2, 3 ]);
    const base64EncodedData = btoa(String.fromCharCode.apply(null, data as unknown as number[]));
    return base64EncodedData;
}
```

### New behavior

> ```csharp
> function someJSMethodReturningAByteArray() {

Author: ```suggestion
```typescript
function someJSMethodReturningAByteArray() {
```

    const data = new Uint8Array([ 1, 2, 3 ]);
    return data;
}
```

## Receiving byte array in JS from .NET

### Old behavior

> ```js
> function ReceivesByteArray(data)
> {

Reviewer 2: ```suggestion
```typescript
function receivesByteArray(data) {
```

    // Previously data was a Base 64 encoded string representing the byte array
}
```

### New behavior

> ```js

Reviewer 2: ```suggestion
```typescript
```

> function ReceivesByteArray(data)
> {

Reviewer 2: ```suggestion
function receivesByteArray(data) {
```

    // 6.0 Preview 6 and beyond, it'll be a Uint8Array (no longer requires processing the Base 64 encoding)
}
```

## Reason for change

Create a more efficient interop mechanism for byte arrays.

## Recommended action

### Receiving byte array in JS from .NET

> If you sent a byte array from .NET:

Reviewer 2: ```suggestion
Consider this .NET interop, where you call into JavaScript passing a byte array:
```

```csharp
var bytes = new byte[] { 1, 5, 7 };

> await _jsRuntime.InvokeVoidAsync("ReceivesByteArray", bytes);

Reviewer 2: ```suggestion
await _jsRuntime.InvokeVoidAsync("receivesByteArray", bytes);
```

```

> Then treat the incoming parameter in JS as a byte array instead of a Base64 encoded string.

Reviewer 2: ```suggestion
In the preceding code example, you'd treat the incoming parameter in JavaScript as a byte array instead of a Base64 encoded string.
```

### Returning byte array from JS to .NET

> If .NET is expecting a `byte[]` JS must provide a `Uint8Array`. Previously, it was possible to provide a Base64 encoded array using `btoa`.

Reviewer: ```suggestion
If .NET expects a `byte[]`, JS must provide a `Uint8Array`. Previously, it was possible to provide a Base64-encoded array using `btoa`.
```

> For example, if you have something like this:

Reviewer 2: ```suggestion
For example, consider the following JavaScript interop call from .NET:
```

> For example, if you have something like this:

Reviewer: ```suggestion
For example, if you have the following code, then you must provide a Uint8Array from JS that's not Base64 encoded:
```

```csharp
var bytes = await _jsRuntime.InvokeAsync<byte[]>("someJSMethodReturningAByteArray");
```

> then you must provide a Uint8Array from JS (must not be Base 64 encoded).

Reviewer 2: ```suggestion
In the preceding code, the `Uint8Array` is returned from JavaScript and must _not_ be Base 64 encoded.
```

> then you must provide a Uint8Array from JS (must not be Base 64 encoded).

Reviewer: ```suggestion
```
